src/db.rs

```diff
@@ -14,25 +15,41 @@ pub struct db_t {
 // and then check it when converting to GoResult manually
 #[repr(C)]
 pub struct DB_vtable {
-    pub read_db: extern "C" fn(*mut db_t, Buffer, *mut Buffer) -> i32,
-    pub write_db: extern "C" fn(*mut db_t, Buffer, Buffer) -> i32,
-    pub remove_db: extern "C" fn(*mut db_t, Buffer) -> i32,
+    pub read_db: extern "C" fn(*mut db_t, *mut gas_meter_t, *mut u64, Buffer, *mut Buffer) -> i32,
+    pub write_db: extern "C" fn(*mut db_t, *mut gas_meter_t, *mut u64, Buffer, Buffer) -> i32,
+    pub remove_db: extern "C" fn(*mut db_t, *mut gas_meter_t, *mut u64, Buffer) -> i32,
     // order -> Ascending = 1, Descending = 2
-    pub scan_db: extern "C" fn(*mut db_t, Buffer, Buffer, i32, *mut GoIter) -> i32,
+    pub scan_db: extern "C" fn(
+        *mut db_t,
+        *mut gas_meter_t,
+        *mut u64,
+        Buffer,
+        Buffer,
+        i32,
+        *mut GoIter,
+    ) -> i32,
 }
 
 #[repr(C)]
 pub struct DB {
+    pub gas_meter: *mut gas_meter_t,
     pub state: *mut db_t,
     pub vtable: DB_vtable,
 }
 
 impl ReadonlyStorage for DB {
-    fn get(&self, key: &[u8]) -> FfiResult<Option<Vec<u8>>> {
+    fn get(&self, key: &[u8]) -> FfiResult<(Option<Vec<u8>>, u64)> {
```

Comment:
this is a bit odd return value, but correct.
i didn't look much at how the cosmwasm_vm code handles it, but i assume like the following.
happy for a link to that (on relevant pr, sorry it's late and I'm lazy)

Reply:
Yeah, the relevant code in cosmwasm_vm is [here](https://github.com/CosmWasm/cosmwasm/pull/379/files#diff-538f23d56c9fe02f65249e0e936ee127R11).
This is really ugly tbh. It was in fact so ugly that even `clippy` complained, and made me refactor the Iterator to define the `StorageIteratorItem` type :/ But, it is logically and structurally correct, and I think anything else would just make it worse in terms of code complexity or usability...